pubsub/publisher.go

```diff
@@ -0,0 +1,81 @@
+package pubsub
+
+import (
+	"cloud.google.com/go/pubsub"
+	"context"
+	"encoding/json"
+	"errors"
+	queuesgo "github.com/merlinapp/queues-go"
+	"reflect"
+)
+
+type publisher struct {
+	topic      *pubsub.Topic
+	objectType reflect.Type
+}
+
+/*
+Creates a new Google's pubsub publisher
+the topic must exists already on the given project
+the objectType interface should be any of the following types, any other type will cause an error returning a nil value
+1. Copy of a structure
+2. No nil pointer of a structure
+3. A map with key string and any value
+*/
+func NewPublisher(project, topic string, objectType interface{}) queuesgo.Publisher {
+	pubsubClient, _ := pubsub.NewClient(context.Background(), project)
+	t := pubsubClient.Topic(topic)
+	if !validateType(objectType) {
+		return nil
+	}
+	return &publisher{
+		topic:      t,
+		objectType: reflect.TypeOf(objectType),
+	}
+}
+
+func (p *publisher) PublishSync(ctx context.Context, event *queuesgo.Event) (string, error) {
+	message, err := p.eventToPubSub(event)
+	if err != nil {
+		return "", err
+	}
+	result := p.topic.Publish(ctx, message)
+	return result.Get(ctx)
+}
+
+func (p *publisher) PublishAsync(ctx context.Context, event *queuesgo.Event) (<-chan struct{}, error) {
+	message, err := p.eventToPubSub(event)
+	if err != nil {
+		return nil, err
+	}
+	result := p.topic.Publish(ctx, message)
+	res := result.Ready()
+	return res, nil
```

Comment:
I'm thinking this bit more in these lines (apologies for any bugs, I wrote it in GitHub haha)

```go
res := make(chan PublicationResult, error)
result := p.topic.Publish(ctx, message)
go(func() {
  s, err := result.Get(ctx)
  p := PublicationResult { s, err  }
  res <- p
})
return result, err
```

Reply:
Done